replication/OrderedReplication.tla

```diff
@@ -0,0 +1,349 @@
+                replicaOffsets[r] + 1
+            ELSE 
+                replicaOffsets[r]
+            ]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, inflightReplicationRequest, inflightReplicationResponse, replicaTokens>>
+
+
+(* Ambry uses a pull model for replication.
+ * In SendReplicationRequest action, a replica asks one of its peers to give the set of data the peer has.
+ * This action just enqueues the request to a peer in "inflightReplicationRequest".
+ * The request contains a replicaToken, which is the next offset of the target replica that target replica needs to send as a response.
+ *)
+SendReplicationRequest(src) ==
+    /\ replicaStates[src] = CanTakeWrite \* We want replica asking for the replication pull to be writable.
+    /\ \E tgt \in Replicas:
+        /\ src # tgt
+        /\ orderedStoredMessagesPerReplica[tgt] # {}
+        /\ inflightReplicationRequest' = [inflightReplicationRequest EXCEPT ![tgt] = inflightReplicationRequest[tgt] \union { ReplicationRequestMessage(src, replicaTokens[src][tgt]) } ]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, inflightReplicationResponse, replicaTokens, orderedStoredMessagesPerReplica, replicaOffsets>>
+
+(* Ambry uses a pull model for replication.
+ * In RecvReplicationRequest action, a replica recieves replication request from a peer.
+ * The request contains a replica token, which is the offset of the message upto which the peer is uptodate.
+ * The recieving replica checks if it has message after the offset in the replica token, and sends the message along with the new token.
+ *)
+RecvReplicationRequest(r) ==
+    /\ inflightReplicationRequest[r] # {}
+    /\ orderedStoredMessagesPerReplica[r] # {} \* there should be something to replicate.
+    /\ \E msg \in inflightReplicationRequest[r]:
+        /\ inflightReplicationResponse' = [inflightReplicationResponse EXCEPT ![msg.src] = inflightReplicationResponse[msg.src] \union ReplicationResponseMessage(r, orderedStoredMessagesPerReplica, msg) ]
+        /\ inflightReplicationRequest' = [inflightReplicationRequest EXCEPT ![r] = inflightReplicationRequest[r] \ {msg}]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, orderedStoredMessagesPerReplica, replicaOffsets, replicaTokens>>
+
+(* Ambry uses a pull model for replication.
+ * In RecvReplicationResponse action, a replica recieves replication response from a peer.
+ * On reciept it checks if each message is present in its local storage log. If not present, then it adds the message, and updates its replica token.
+ *)
+RecvReplicationResponse(r) ==
+    /\ inflightReplicationResponse[r] # {}
+    /\ \E msg \in inflightReplicationResponse[r]:
+        /\ \E message \in msg.messages:
+            /\ IF message.blobmessage = UPDATE_TTL THEN
```

Comment:
Maybe pull some of these out to UPDATE_TTL/PUT/DELETE just for readabilities sake if possible?